chainerx_cc/chainerx/cuda/reduce.cuh

```diff
@@ -188,5 +252,44 @@ void Reduce(const Array& in, const Axes& axis, const Array& out, ReductionImpl&&
             MakeReductionKernelArg<In, Out>(arg), out_block_size, reduce_block_size, impl);
 }
 
+template <typename In, typename Out, typename ReductionImpl>
+void Scan(const Array& in, int8_t axis, const Array& out, ReductionImpl&& impl) {
+    if (out.GetTotalSize() == 0) {
+        return;
+    }
+
+    ReductionArg arg{in, Axes{axis}, out};
+    int64_t reduce_len = in.shape()[axis];
+
+    // TODO(niboshi): Calculate kMaxBlockSize per device
+    std::lock_guard<std::mutex> lock{*cuda_internal::g_mutex};
+    static const int64_t kMaxBlockSize = std::min(
+            reduce_detail::kMaxReductionBlockSize,
+            CudaOccupancyMaxPotentialBlockSize(&reduce_detail::ReductionKernel<In, Out, ReductionImpl>).block_size);
+
+    int64_t reduce_total_size_pow2 = reduce_detail::RoundUpToPowerOf2(std::max(int64_t{1}, reduce_len));
+
+    int64_t reduce_block_size = std::min(kMaxBlockSize, reduce_total_size_pow2);
+    int64_t out_block_size = kMaxBlockSize / reduce_block_size;
+    int64_t out_block_num = (arg.in_shape().GetTotalSize() / reduce_len + out_block_size - 1) / out_block_size;
+
+    int64_t block_size = kMaxBlockSize;
+    int64_t grid_size = std::min(reduce_detail::kMaxGridSize, out_block_num);
+    int64_t shared_mem_size = sizeof(decltype(impl.Identity())) * block_size;
+
+#ifdef NDEBUG  // Optimize only in Release build to save time on development
+    // TODO(sonots): Reconsider the number of statically-optimized kernels in terms of speed and binary size trade-offs.
+    // Currently, only contiguous output arrays are optimized.
+    if (arg.in_strides().ndim() == 1 && arg.out_strides().ndim() == 1) {
+        reduce_detail::ScanKernel<<<grid_size, block_size, shared_mem_size>>>(
+                MakeReductionKernelArg<In, Out, 1, 1>(arg), out_block_size, reduce_block_size, impl, reduce_len);
```

Comment:
here arg is called `reduce_len` while in the kernel code is `reduction_dim`
bit confusing.

Reply:
I agree with your suggestion, and I want to apply a similar fix also in `native/reduce.h`. Can I do it in this PR?